src/Agent.Worker/TfManager.cs

```diff
@@ -0,0 +1,131 @@
+                    executionContext.Error(ex);
+                }
+            }
+            else
+            {
+                executionContext.Debug($"tf-legacy download already exists at {tfLegacyExternalsPath}.");
+            }
+
+            string vstsomLegacyExternalsPath = Path.Combine(externalsPath, "vstsom-legacy");
+
+            if (!Directory.Exists(vstsomLegacyExternalsPath))
+            {
+                const string vstsomDownloadUrl = "https://vstsagenttools.blob.core.windows.net/tools/vstsom/m122_887c6659/vstsom.zip";
+
+                string tempVstsomDirectory = Path.Combine(externalsPath, "vstsom_download_temp");
+                Directory.CreateDirectory(tempVstsomDirectory);
+                string downloadVstsomPath = Path.ChangeExtension(Path.Combine(tempVstsomDirectory, "vstsomdownload"), ".completed");
+
+                if (!File.Exists(downloadVstsomPath))
+                {
+                    await DownloadAsync(downloadVstsomPath, vstsomDownloadUrl, executionContext);
+                }
+                else
+                {
+                    executionContext.Debug($"vstsom is already downloaded to {downloadVstsomPath}.");
+                }
+
+                try
+                {
+                    ZipFile.ExtractToDirectory(downloadVstsomPath, vstsomLegacyExternalsPath);
+                    File.WriteAllText(downloadVstsomPath, DateTime.UtcNow.ToString());
+                    executionContext.Debug("vstsom-legacy has been extracted and cleaned up");
+                }
+                catch (Exception ex)
+                {
+                    executionContext.Error(ex);
+                }
+            }
+            else
+            {
+                executionContext.Debug($"vstsom-legacy download already exists at {vstsomLegacyExternalsPath}.");
+            }
+        }
+
+        private static async Task DownloadAsync(string downloadPath, string blobUrl, IExecutionContext executionContext)
+        {
+            int retryCount = 0;
+            const int retryLimit = 3;
+            const int timeout = 180;
+            const int defaultFileStreamBufferSize = 4096;
+            const int retryDelay = 10000;
+
+            while (retryCount < retryLimit)
+            {
+                using CancellationTokenSource downloadCts = new(TimeSpan.FromSeconds(timeout));
+                using CancellationTokenSource linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(downloadCts.Token, executionContext.CancellationToken);
+                CancellationToken cancellationToken = linkedTokenSource.Token;
+
+                try
+                {
+                    using HttpClient httpClient = new();
+                    using Stream stream = await httpClient.GetStreamAsync(blobUrl, cancellationToken);
+                    using FileStream fs = new(downloadPath, FileMode.Create, FileAccess.Write, FileShare.None, bufferSize: defaultFileStreamBufferSize, useAsync: true);
```

Comment:
Should we do a cleanup of a destination file if all attempts were unsuccessful?
If after `retryLimit` times the source file still is not loaded then there will be empty file left created during `fs` initialization.
The next time in the `DownloadTfLegacyAsync` your code will consider that file already exists but it might be corrupted

Reply:
Good point, thanks. I added cleanup for temporary download files